grails-app/controllers/au/org/ala/LoginInterceptor.groovy

```diff
@@ -0,0 +1,105 @@
+        if (grailsApplication.config.security.cas.disableCAS.toBoolean() || grailsApplication.config.security.cas.bypass.toBoolean()) {
+            return true
+        }
+
+        def controller = grailsApplication.getArtefactByLogicalPropertyName("Controller", controllerName)
+        Class controllerClass = controller?.clazz
+        def method = controllerClass?.getMethod(actionName ?: "index", [] as Class[])
+
+        def role
+        def securityCheck = false // if need to do security check
+
+        if (method?.isAnnotationPresent(RequireAdmin) || controllerClass?.isAnnotationPresent(RequireAdmin)) {
+            role= grailsApplication.config.auth.admin_role //recommended: ROLE_ADMIN
+            securityCheck = true
+        }else if (method?.isAnnotationPresent(RequireLogin) || controllerClass?.isAnnotationPresent(RequireLogin)) {
+            RequireLogin requireLogin = method.getAnnotation(RequireLogin.class)
+            role=requireLogin.role()
+            securityCheck = true
+        }
+        //Should be the last step
+        if(method?.isAnnotationPresent(SkipSecurityCheck)){
+            securityCheck = false
+        }
+
+        if(securityCheck){
+            String apikey = getApiKey() //collect apikey from header, param and body
+
+            if(serviceAuthService.isValid(apikey)){
+                true
+            }else{
+                if (Strings.isNullOrEmpty(role)) {
+                    //Request login only
+                    if (!serviceAuthService.isLoggedIn()) {
+                        //TODO check type of request, determine whether returns JSON or redirect to login page
+                        redirect(url: grailsApplication.config.security.cas.loginUrl + "?service=" +
+                                grailsApplication.config.security.cas.appServerName + request.forwardURI + (request.queryString ? '?' + request.queryString : ''))
+                        return false
+                    }
+                } else {
+                    if (!serviceAuthService.isRoleOf(role)) {
+                        //TODO check type of request, determine whether returns JSON or redirect to login page
+                        render text: "Forbidden. " + role + " required", status: STATUS_UNAUTHORISED
+                        return false
+                    }
+                }
+            }
+        }
+        true
+    }
+
+    boolean after() { true }
+
+    void afterView() {
+        // no-op
+    }
+
+    /**
+     * ALA uses 'apiKey' as standard
+     * @return
+     */
+    private getApiKey(){
+        String apikey
+        if (request.getHeader(API_KEY_HEADER_NAME) != null) {
+            //Standard way
+            apikey = request.getHeader(API_KEY_HEADER_NAME)
+        } else if (params.containsKey("api_key")) {
+            // backward compatible
+            // Check if params contains api_key
+            apikey = params.api_key
+        } else if (request.JSON?.api_key != null) {
```

Comment:
`request.JSON` opens the request inputStream, preventing the controller from doing so. 

It is acceptable to use only when the controller method also makes use of `request.JSON`. e.g. add as an annotation option that is disabled by default if this is required.

Reply:
I ran some tests,  the controller still captures request.JSON
![Screen Shot 2020-12-02 at 3 34 07 pm](https://user-images.githubusercontent.com/31238531/100828770-f9393b80-34b3-11eb-92b4-45e9e74f0a61.png)

Reply:
Removed